funsor/adjoint.py

```diff
@@ -149,3 +154,112 @@ def adjoint_contract(adj_redop, adj_binop, out_adj, sum_op, prod_op, reduced_var
+
+    # ones for things that were sliced away
+    ones_like_arg = Tensor(torch.full_like(arg.data, ops.UNITS[adj_binop]),
+                           arg.inputs.copy(), arg.output.dtype)
+    arg_adj = _scatter(arg_adj, ones_like_arg, slices)
+
+    return {arg: arg_adj}
+
+
+def _scatter(src, res, subs):
+    # inverse of advanced indexing
+    assert isinstance(res, Tensor)
+    assert isinstance(src, Tensor)
+    # TODO check types of subs, in case some logic from eager_subs was accidentally left out?
+
+    # use advanced indexing logic copied from Tensor.eager_subs:
+
+    # materialize after checking for renaming case
+    subs = OrderedDict((k, materialize(v)) for k, v in subs)
+
+    # Compute result shapes.
+    inputs = OrderedDict()
+    for k, domain in res.inputs.items():
+        inputs[k] = domain
+
+    # Construct a dict with each input's positional dim,
+    # counting from the right so as to support broadcasting.
+    total_size = len(inputs) + len(res.output.shape)  # Assumes only scalar indices.
+    new_dims = {}
+    for k, domain in inputs.items():
+        assert not domain.shape
+        new_dims[k] = len(new_dims) - total_size
+
+    # Use advanced indexing to construct a simultaneous substitution.
+    index = []
+    for k, domain in res.inputs.items():
+        if k in subs:
+            v = subs.get(k)
+            if isinstance(v, Number):
+                index.append(int(v.data))
+            else:
+                # Permute and expand v.data to end up at new_dims.
+                assert isinstance(v, Tensor)
+                v = v.align(tuple(k2 for k2 in inputs if k2 in v.inputs))
+                assert isinstance(v, Tensor)
+                v_shape = [1] * total_size
+                for k2, size in zip(v.inputs, v.data.shape):
+                    v_shape[new_dims[k2]] = size
+                index.append(v.data.reshape(tuple(v_shape)))
+        else:
+            # Construct a [:] slice for this preserved input.
+            offset_from_right = -1 - new_dims[k]
+            index.append(torch.arange(domain.dtype).reshape(
+                (-1,) + (1,) * offset_from_right))
+
+    # Construct a [:] slice for the output.
+    for i, size in enumerate(res.output.shape):
+        offset_from_right = len(res.output.shape) - i - 1
+        index.append(torch.arange(size).reshape(
+            (-1,) + (1,) * offset_from_right))
+
+    # the only difference from Tensor.eager_subs is here:
+    # instead of indexing the rhs (lhs = rhs[index]), we index the lhs (lhs[index] = rhs)
+
+    # unsqueeze to make broadcasting work
+    src_inputs, src_data = src.inputs.copy(), src.data
+    for k, v in res.inputs.items():
+        if k not in src.inputs and isinstance(subs[k], Number):
+            src_inputs[k] = bint(1)
+            src_data = src_data.unsqueeze(-1)
+    src = Tensor(src_data, src_inputs, src.output.dtype).align(tuple(res.inputs.keys()))
+
+    data = res.data
+    data[tuple(index)] = src.data
```

Comment:
Here we reuse the indexing logic from `Tensor.eager_subs` to do a fancy broadcasted assignment instead of a fancy broadcasted selection.